lib/revolution.rb

```
@@ -1,9 +1,135 @@
+require 'revolution/order'
+require 'revolution/recipes'
+require 'revolution/exceptions'
+require 'thor'
 
 module Revolution
-  def self.run
-    puts 'Running!'
+  # TODO: separate CLI class from main Revolution logic
+  class CLI < Thor
+    desc 'all', 'Start Revolution'
+    option :recipe_root, required: true
+    def all
+      puts 'Doing all the things!'
+      diff
+      build
+      sign
+      deploy
+    rescue Error::ExecutionFailure
+      exit 1
+    end
+
+    desc 'diff', 'Check for packages that have changed'
+    option :recipe_root, required: true
+    def diff
+      puts 'Checking for package updates...'
+      # TODO
+    end
+
+    desc 'order', 'Get build order based on dependencies'
+    option :recipe_root, required: true
+    def order
+      puts 'Getting build order...'
+      return StandardError unless valid_dir?(options.recipe_root)
+      recipes        = Recipes.load_recipes(options.recipe_root)
+      sorted_recipes = Order.resolve_build_order(recipes)
+      output_build_order(sorted_recipes)
+      sorted_recipes
+    rescue Error::ExecutionFailure
+      exit 1
+    end
+
+    desc 'build', 'Build packages in order according to dependencies'
+    option :recipe_root, required: true
+    # TODO: add --all option, package_name parameter
+    def build
+      puts 'Building packages in order...'
+      sorted_recipes = order # gets options.recipe_root implicitly
+      built_packages = []
+      sorted_recipes.each do |recipe|
+        puts "Building package: #{recipe.package_name}"
+        package_path = options.recipe_root + recipe.package_name
+        Build.clean(package_path)
+        Build.build_package(package_path)
+        Build.clean(package_path)
+        built_packages.push(recipe.package_name)
+      end
+      puts 'Successfully built the following packages:'
+      built_packages.each do |pkg_name|
+        puts "    #{pkg_name}"
+      end
+      'success'
+    rescue Error::ExecutionFailure
+      exit 1
```

Wont this swallow the error and not bubble up?

---

Also, you're missing a begin aren't you?

---

RuboCop calls a `begin` there redundant.